kernel/src/main/java/org/kframework/parser/inner/disambiguation/PriorityVisitor.java

```diff
@@ -35,22 +38,55 @@ public PriorityVisitor(POSet<Tag> priorities, Set<Tuple2<Tag, Tag>> leftAssoc, S
     @Override
     public Either<java.util.Set<KEMException>, Term> apply(TermCons tc) {
         assert tc.production() != null : this.getClass() + ":" + " production not found." + tc;
-        if (!tc.production().isSyntacticSubsort() && !tc.production().att().contains(Att.BRACKET())) {
+        if (!tc.production().isSyntacticSubsort()) {
             // match only on the outermost elements
-            if (tc.production().items().apply(0) instanceof NonTerminal) {
-                Either<java.util.Set<KEMException>, Term> rez =
-                        new PriorityVisitor2(tc, PriorityVisitor2.Side.LEFT, priorities, leftAssoc, rightAssoc).apply(tc.get(0));
-                if (rez.isLeft())
-                    return rez;
-                tc = tc.with(0, rez.right().get());
-            }
-            if (tc.production().items().apply(tc.production().items().size() - 1) instanceof NonTerminal) {
-                int last = tc.items().size() - 1;
-                Either<java.util.Set<KEMException>, Term> rez =
-                        new PriorityVisitor2(tc, PriorityVisitor2.Side.RIGHT, priorities, leftAssoc, rightAssoc).apply(tc.get(last));
-                if (rez.isLeft())
-                    return rez;
-                tc = tc.with(last, rez.right().get());
+            if (tc.production().att().contains("applyPriority")) {
+              String[] pieces = StringUtil.splitOneDimensionalAtt(tc.production().att().get("applyPriority"));
```

Comment:
Doing this at every step might be slow.

Reply:
This is not happening inside a loop. The only way we could precompute this value would be to do some kind of preprocessing pass in another visitor and then store the values in structured form as attributes and read them here. That would be a lot more complicated to code, because currently we don't have any kind of semantic analysis phase of compilation. Also, you will note we do the same thing with other attributes later on in the compilation process in places where it doesn't seem to have much effect on performance, so I'm not really worried about it. Let's optimize based on data, not based on what we assume will be slow.